### ISSUE TITLE
ROP builder wrong behavior for multiple binaries

### ISSUE THREAD

**Issue by quantatic** [Gallopsled/pwntools#1329]
When building a ROP chain with multiple ELFs, there seems to be a bug.

I'm getting wildly varying behavior depending on the order in which I load the ELFs, which definitely should not be happening, gadgets using the wrong base address, and other issues.

The easiest way to demonstrate the problem is the following:

```python
from pwn import *
context.arch = "amd64"

e1 = ELF.from_assembly("pop rdi; ret")
e2 = ELF.from_assembly("pop rdi; ret; pop rax; ret; pop rdx; ret")

rop1 = ROP([e1, e2])
rop2 = ROP([e2, e1])

print(len(rop1.gadgets)) #should print 2
print(len(rop2.gadgets)) #should print 4
```

rop1 and rop2 will have a different number of gadgets, even though they are both built from the exact same ELFs, which shouldn't be happening as it should be finding gadgets from both passed ELFs.

**Comment by zachriggle** [comment]
This is exactly what I see when I test this on my machine, if I copy-paste the script.
```py
# issue1329.py
from pwn import *
context.arch = "amd64"

e1 = ELF.from_assembly("pop rdi; ret")
e2 = ELF.from_assembly("pop rdi; ret; pop rax; ret; pop rdx; ret")

rop1 = ROP([e1, e2])
rop2 = ROP([e2, e1])

print(len(rop1.gadgets)) #should print 2
print(len(rop2.gadgets)) #should print 4
```
```
$ python issue1329.py LOG_LEVEL=ERROR
2
4
```

**Comment by zachriggle** [comment]
I misunderstood the original problem, and have reopened this.

**Comment by zachriggle** [comment]
This might be due to sorting of hashes in cache file names, and also due to caching not being deterministic when passing in [a,b] vs [b,a].

**Comment by peace-maker** [comment]
Fixed in #1369
